src/membership/cut_detector.rs

```diff
@@ -0,0 +1,162 @@
+use crate::{
+    common::Endpoint,
+    transport::proto::{AlertMessage, EdgeStatus},
+};
+
+use std::collection::{HashMap, HashSet};
+
+const NUM_MIN: usize = 3;
+
+pub struct MultiNodeCutDetector {
+    /// Number of observers per subject and vice versa
+    num: usize,
+    /// High water mark
+    high: usize,
+    /// Low water mark
+    low: usize,
+    proposal_count: usize,
+    updates_in_progress: usize,
+    reports_per_host: HashMap<Endpoint, HashMap<usize, Endpoint>>,
+    proposal: Vec<Endpoint>,
+    pre_proposal: HashSet<Endpoint>,
+    seen_down_link_events: bool,
+}
+
+impl MultiNodeCutDetector {
+    pub fn new(num: usize, high: usize, low: usize) -> Self {
+        if high > num || low > high || num < NUM_MIN || low <= 0 || high <= 0 {
+            panic!("Invalid arguments passed to cut detector")
+        }
+
+        Self {
+            num,
+            high,
+            low,
+            proposal_count: 0,
+            updates_in_progress: 0,
+            reports_per_host: HashMap::new(),
+            proposal: Vec::new(),
+            pre_proposal: HashSet::new(),
+            seen_down_link_events: false,
+        }
+    }
+
+    fn aggregate(message: AlertMessage) -> Vec<Endpoint> {
```

Comment:
I wonder if we can make this spit out another iterator instead of a vec?

Reply:
That might be a good idea. I'll look into it.